### ISSUE TITLE
Consider using UMP for clap_event_midi (similar to clap_event_midi2)

### ISSUE THREAD

**Issue by Bremmers** [free-audio/clap#48]
Apple has already deprecated all 'old' MIDI APIs in favor of UMP versions. Microsoft will likely use UMP for MIDI 1.0 too when a MIDI 2.0 API appears. That's simply the way the new MIDI is designed, and the way MIDI 1.0 will live on in a world that moves to MIDI 2.0. This will make clap_event_midi old fashioned pretty soon.

It would make sense to change clap_event_midi to use UMP instead of traditional 2/3 byte messages. Two helper function can be added to convert between traditional and UMP MIDI.

The dilemma is that everybody will want to use UMP for MIDI 1.0 two years from now, but most people don't know that yet ;-)

PS: Many issues are solved, and it looks good. Great work! :-)

**Comment by abique** [comment]
Why not simply use midi2 then?

**Comment by Bremmers** [comment]
UMP and the MIDI 2.0 protocol aren't the same thing.

UMP is the message type, where each message is 1..4 32 bits words.

The MIDI 2.0 protocol is the MIDI protocol with higher resolution messages, and some new messages (most notably per-note controllers).

UMP messages can contain either the MIDI 1.0 or MIDI 2.0 protocol. The MIDI 1.0 UMP messages are basically the 'old' midi message bytes in reversed order, with a one byte header.

I think CLAP_EVENT_MIDI2 means "the MIDI 2.0 protocol". If you intend it to mean "UMP messages" there's a problem as the host currently can't detect whether the plugin supports the MIDI 1.0 or MIDI 2.0 protocol.

**Comment by Bremmers** [comment]
How about this?

typedef struct clap_event_midi {
   clap_event_header_t header;

   uint16_t port_index;
   uint16_t reserved;      //4 byte-aligns the bytes below
   //don't forget to set messagetype, this is what makes the four bytes an UMP message,
   //this is important for future compatibility
   uint8_t data2;          //third MIDI data byte
   uint8_t data1;          //second MIDI data byte
   uint8_t data0;          //first MIDI data byte (status)
   uint8_t messagetype;    //set to 0x20 if data0 < 0xf0, set to 0x10 if data0 > 0xf0
} clap_event_midi_t;

A plugin can just ignore messagetype and use the 3 data bytes like it used to.

A host can set the 3 data bytes like it used to, and in addition it must set messagetype.
Or, if the host is UMP aware already it can just write a uint32 over the four bytes.

Now we're in sync with the MIDI APIs from Apple (and all other OSes that add support for UMP/MIDI 2.0). The 4 bytes can be turned into a single "uint32_t umpdata;" field in the future. And plugin developers won't even notice this is different from what they're used to :-)

**Comment by abique** [comment]
Is your layout endian independent?

It'd be like this then?
```C
typedef struct clap_event_midi {
   clap_event_header_t header;

   uint16_t port_index;
   union {
      uint32_t ump_data;
      struct {
         uint8_t data2;       // third MIDI data byte
         uint8_t data1;       // second MIDI data byte
         uint8_t data0;       // first MIDI data byte (status)
         uint8_t messagetype; // set to 0x20 if data0 < 0xf0, set to 0x10 if data0 > 0xf0
      };
   };
} clap_event_midi_t;
``` ?

**Comment by abique** [comment]
I gave a good thought about it.
1. UMP goes beyond what midi 1.0 is, and if `message_type = f(data0)` then `message_type` is not really useful
2. Some will use UMP beyond its scope and maybe produce non MIDI 1.0 data, which will force everyone to support it
3. As we work with `uint32_t` there is a question of byte order. After a quick search, I haven't found a standard way to figure it out via macros.
   - re-ordering the fields will be difficult
      - it would make the struct look more complicated than it needs to be
   - the simple approach would be to only work with `uint32_t` then, in native endian
      - even if you work with `uint32_t`, at some point you'll need to deconstruct it into `status, data0, data1`
4. It is easy to convert `uint8_t data[3];` to/from UMP, so no big deal
5. UMP requires to know how to work with MIDI 1.0 data, but working with MIDI 1.0 data doesn't require to know UMP
   
In conclusion, I'm not convinced by using UMP for MIDI 1.0 data.
